### PR TITLE
Refactors firestacks into status effects

### PR DESCRIPTION
**Pull request by SmArtKar**
## About The Pull Request

This PR refactors firestacks into two status effects: fire_stacks, which behave like normal firestacks you have right now, and wet_stacks, which are your negative fire stacks right now. This allows for custom fires with custom behaviors and icons to be made.

Some fire related is moved away from species(what the fuck was it even doing there) into these as well.
Oh and I fixed the bug where monkeys on fire had a human fire overlay, why wasn't this fixed already, it's like ancient.

Also changed some related proc names to be snake_case like everything should be.

## Why It's Good For The Game

This allows for custom fire types with custom behaviours, like freezing freon fire or radioactive tritium fire. Removing vars from living and moving them to status effects for modularity is also good.
Nothing to argue about since there's nothing player-facing

## Changelog

:cl:
code: Changed some fire-related procs to be snake_case
refactor: Refactored firestacks into status effects
fix: Monkeys no longer have human fire overlay when on fire and instead use the one they're supposed to use.
/:cl:

### CODE REVIEW

File: code/datums/status_effects/debuffs/debuffs.dm [hunk 1]
```diff
@@ -1322,3 +1322,281 @@
...
+
+/datum/status_effect/fire_handler
+	duration = -1
+	alert_type = null
+	status_type = STATUS_EFFECT_MULTIPLE //Custom code
+	on_remove_on_mob_delete = TRUE
+	var/stacks
+	var/stack_limit = 20
+	var/our_type
+	var/enemy_type
```
Comment (SuperNovaa41) [inline code/datums/status_effects/debuffs/debuffs.dm:1334]: me when n o docs

File: code/datums/status_effects/debuffs/debuffs.dm [hunk 2]
```diff
@@ -1322,3 +1322,281 @@
...
+	id = "fire_stacks" //fire_stacks and wet_stacks should have different IDs or else has_status_effect won't work
+
+	our_type = /datum/status_effect/fire_handler/fire_stacks
+	enemy_type = /datum/status_effect/fire_handler/wet_stacks
+
+	var/on_fire = FALSE
+	var/fire_alert_type = /atom/movable/screen/alert/fire
+	var/datum/weakref/firelight_ref
+	var/firelight_type = /obj/effect/dummy/lighting_obj/moblight/fire
+	var/last_icon_state
```
Comment (SuperNovaa41) [inline code/datums/status_effects/debuffs/debuffs.dm:1401]: autodocs

File: code/datums/status_effects/debuffs/debuffs.dm [hunk 3]
```diff
@@ -1322,3 +1322,281 @@
...
+		leg_clothes = victim.shoes
+	else if(victim.wear_suit && ((victim.wear_suit.body_parts_covered & FEET) || (victim.wear_suit.body_parts_covered & LEGS)))
+		leg_clothes = victim.wear_suit
+	else if(victim.w_uniform && ((victim.w_uniform.body_parts_covered & FEET) || (victim.w_uniform.body_parts_covered & LEGS)))
+		leg_clothes = victim.w_uniform
+	if(leg_clothes)
+		burning_items |= leg_clothes
+
+	for(var/X in burning_items)
+		var/obj/item/I = X
```
Comment (SuperNovaa41) [inline code/datums/status_effects/debuffs/debuffs.dm:1503]: bad var names

**Comment by DAKKA-WAAAGH** [comment]
Guh

File: code/modules/mob/living/living.dm [hunk 4]
```diff
@@ -1492,6 +1492,8 @@
 	if(wet_status)
 		return -wet_status.stacks
 
+	return 0
```
Comment (Ghilker) [inline code/modules/mob/living/living.dm:1495]: return FALSE

File: code/datums/status_effects/debuffs/debuffs.dm [hunk 5]
```diff
@@ -1322,3 +1322,289 @@
...
+				if(forced)
+					stacks = min(stacks, stack_limit)
+				else
+					stacks = min(stacks + our_status.stacks, stack_limit)
+				on_reapplying_stacks(our_status)
+				owner.remove_status_effect(our_type)
+
+	return ..()
+
+/datum/status_effect/fire_handler/proc/on_reapplying_stacks(datum/status_effect/fire_handler/our_status)
```
Comment (MrMelbert) [inline code/datums/status_effects/debuffs/debuffs.dm:1391]: Your procs need auto docs

File: code/datums/status_effects/debuffs/debuffs.dm [hunk 6]
```diff
@@ -1322,3 +1322,289 @@
...
+	/// A weakref to the mob light emitter
+	var/datum/weakref/firelight_ref
+	/// Type of mob light emitter we use when on fire
+	var/firelight_type = /obj/effect/dummy/lighting_obj/moblight/fire
+	/// Stores current fire overlay icon state, for optimisation purposes
+	var/last_icon_state
+
+/datum/status_effect/fire_handler/fire_stacks/tick(delta_time, times_fired)
+	if(stacks <= 0)
+		owner.remove_status_effect(type)
```
Comment (MrMelbert) [inline code/datums/status_effects/debuffs/debuffs.dm:1414]:
```suggestion
		qdel(src)
```
You can just `qdel(src)` a status effect to self-remove it.

File: code/datums/status_effects/debuffs/debuffs.dm [hunk 7]
```diff
@@ -1322,3 +1322,289 @@
...
+	/// Type for our alert
+	var/fire_alert_type = /atom/movable/screen/alert/fire
+	/// A weakref to the mob light emitter
+	var/datum/weakref/firelight_ref
+	/// Type of mob light emitter we use when on fire
+	var/firelight_type = /obj/effect/dummy/lighting_obj/moblight/fire
+	/// Stores current fire overlay icon state, for optimisation purposes
+	var/last_icon_state
+
+/datum/status_effect/fire_handler/fire_stacks/tick(delta_time, times_fired)
```
Comment (MrMelbert) [inline code/datums/status_effects/debuffs/debuffs.dm:1412]: You're changing fire from ticking every 2 seconds (`Life()` ticks roughly every two seconds) to ticking every 1 second. You need to either set the `tick_interval` to `2 SECONDS` or half all the values associated.

File: code/datums/status_effects/debuffs/debuffs.dm [hunk 8]
```diff
@@ -1322,3 +1322,289 @@
...
+	stacks = new_stacks
+
+	var/datum/status_effect/fire_handler/our_status = owner.has_status_effect(our_type)
+	var/datum/status_effect/fire_handler/enemy_status = owner.has_status_effect(enemy_type)
+
+	if(enemy_status)
+		stacks -= enemy_status.stacks
+		enemy_status.stacks -= new_stacks
+		if(enemy_status.stacks <= 0)
+			owner.remove_status_effect(enemy_type)
```
Comment (MrMelbert) [inline code/datums/status_effects/debuffs/debuffs.dm:1351]: You can just `qdel(enemy_status)` since you already have a reference to it.

File: code/datums/status_effects/debuffs/debuffs.dm [hunk 9]
```diff
@@ -1322,3 +1322,289 @@
...
+	/// What status effect type do we merge with
+	var/our_type
+	/// What status effect type do we remove uppon being applied
+	var/enemy_type
+
+/datum/status_effect/fire_handler/on_creation(mob/living/new_owner, new_stacks, override_type = NO_FIRE_OVERRIDE, forced = FALSE)
+	owner = new_owner
+	stacks = new_stacks
+
+	var/datum/status_effect/fire_handler/our_status = owner.has_status_effect(our_type)
```
Comment (MrMelbert) [inline code/datums/status_effects/debuffs/debuffs.dm:1344]: Seems like in most cases `our_status` = `src`, so why bother grabbing it?

File: code/modules/mob/living/carbon/examine.dm [hunk 10]
```diff
@@ -96,9 +96,9 @@
 	if(HAS_TRAIT(src, TRAIT_DUMB))
 		msg += "[t_He] seem[p_s()] to be clumsy and unable to think.\n"
 
-	if(fire_stacks > 0)
+	if(get_fire_stacks() > 0)
 		msg += "[t_He] [t_is] covered in something flammable.\n"
-	if(fire_stacks < 0)
+	if(get_fire_stacks() < 0)
```
Comment (MrMelbert) [inline code/modules/mob/living/carbon/examine.dm:101]:
```suggestion
	if(has_status_effect(/datum/status_effect/fire_handler/wet_stacks)
```

File: code/modules/mob/living/carbon/human/examine.dm [hunk 11]
```diff
@@ -216,9 +216,9 @@
 				msg += "<b>[t_He] [t_has] severe cellular damage!</b>\n"
 
 
-	if(fire_stacks > 0)
+	if(get_fire_stacks() > 0)
 		msg += "[t_He] [t_is] covered in something flammable.\n"
-	if(fire_stacks < 0)
+	if(get_fire_stacks() < 0)
```
Comment (MrMelbert) [inline code/modules/mob/living/carbon/human/examine.dm:221]:
```suggestion
	if(has_status_effect(/datum/status_effect/fire_handler/wet_stacks)
```

File: code/modules/mob/living/silicon/robot/examine.dm [hunk 12]
```diff
@@ -20,9 +20,9 @@
 			. += span_warning("<B>It looks severely burnt and heat-warped!</B>")
 	if (health < -maxHealth*0.5)
 		. += span_warning("It looks barely operational.")
-	if (fire_stacks < 0)
+	if (get_fire_stacks() < 0)
 		. += span_warning("It's covered in water.")
-	else if (fire_stacks > 0)
+	else if (get_fire_stacks() > 0)
```
Comment (MrMelbert) [inline code/modules/mob/living/silicon/robot/examine.dm:25]:
```suggestion
	if(has_status_effect(/datum/status_effect/fire_handler/wet_stacks)
```

File: code/modules/mob/living/carbon/examine.dm (hunk 10 above)
Comment (MrMelbert) [inline code/modules/mob/living/carbon/examine.dm:101]: Also, IMO you should handle these via the status effect's `examine_text` instead. Though I refactored it in #66340 so you may want to wait for that, since it'll make it easier.

File: code/datums/status_effects/debuffs/debuffs.dm [hunk 13]
```diff
@@ -1322,3 +1322,289 @@
...
+	if(iscyborg(owner))
+		stacks -= 0.5 * delta_time
+
+	if(stacks <= 0)
+		owner.remove_status_effect(type)
+		return TRUE
+
+	var/datum/gas_mixture/air = owner.loc.return_air()
+	if(!air.gases[/datum/gas/oxygen] || air.gases[/datum/gas/oxygen][MOLES] < 1)
+		owner.remove_status_effect(type)
```
Comment (MrMelbert) [inline code/datums/status_effects/debuffs/debuffs.dm:1431]:
```suggestion
		qdel(src)
```

**Review by MrMelbert — CHANGES_REQUESTED** [review]

File: code/__DEFINES/ai.dm [hunk 14]
```diff
@@ -36,7 +36,7 @@
 /// probability that the pawn should try resisting out of restraints
 #define RESIST_SUBTREE_PROB 50
 ///macro for whether it's appropriate to resist right now, used by resist subtree
-#define SHOULD_RESIST(source) (source.on_fire || source.buckled || HAS_TRAIT(source, TRAIT_RESTRAINED) || (source.pulledby && source.pulledby.grab_state > GRAB_PASSIVE))
+#define SHOULD_RESIST(source) (source.is_on_fire() || source.buckled || HAS_TRAIT(source, TRAIT_RESTRAINED) || (source.pulledby && source.pulledby.grab_state > GRAB_PASSIVE))
```
Comment (LemonInTheDark) [inline code/__DEFINES/ai.dm:39]: how often are is_on_fire and get_fire_stacks called? is there any reason they can't be cached, as they are now?

File: code/datums/status_effects/debuffs/fire_stacks.dm [hunk 15]
```diff
@@ -10,8 +10,14 @@
 	var/stack_limit = 20
 	/// What status effect type do we remove uppon being applied
 	var/enemy_type
+	/// For how much firestacks does one our stack could
```
Comment (LemonInTheDark) [inline code/datums/status_effects/debuffs/fire_stacks.dm:13]: this seems kinda weird

File: code/datums/status_effects/debuffs/fire_stacks.dm [hunk 16]
```diff
@@ -72,6 +82,13 @@
 
 /datum/status_effect/fire_handler/proc/on_reapplying_stacks(datum/status_effect/fire_handler/our_status)
 
+/**
+ * Recalculates owner's cached firestacks
+ */
+
+/datum/status_effect/fire_handler/proc/recalculate_firestacks()
+	owner.fire_stacks = max(0, stacks) * stack_modifier
```
Comment (LemonInTheDark) [inline code/datums/status_effects/debuffs/fire_stacks.dm:90]: It might be best to make an internal setter for fire_stacks, to avoid someone forgetting about it

File: code/datums/status_effects/debuffs/fire_stacks.dm [hunk 17]
```diff
@@ -0,0 +1,373 @@
...
+/**
+ * Refresher for mob's fire_stacks
+ */
+
+/datum/status_effect/fire_handler/proc/cache_stacks()
+	owner.fire_stacks = 0
+	owner.on_fire = FALSE
+	owner.clear_alert(ALERT_FIRE)
+	for(var/fire_type in subtypesof(/datum/status_effect/fire_handler))
+		var/datum/status_effect/fire_handler/possible_fire = owner.has_status_effect(fire_type)
```
Comment (LemonInTheDark) [inline code/datums/status_effects/debuffs/fire_stacks.dm:101]: You should instead do 
```suggestion
	for(var/datum/status_effect/fire_handler/possible_fire in owner.status_effects)
```
It scales with active effects instead of subtypes, but has_status_effect already does this iteration, so this is gonna be cheaper.

image

File: code/datums/status_effects/debuffs/fire_stacks.dm [hunk 18]
```diff
@@ -0,0 +1,373 @@
...
+		if(possible_fire)
+			owner.fire_stacks += possible_fire.stacks * possible_fire.stack_modifier
+			if(istype(possible_fire))
+				var/datum/status_effect/fire_handler/fire_stacks/our_fire = possible_fire
+				if(our_fire.on_fire)
+					owner.on_fire = TRUE
+					owner.throw_alert(ALERT_FIRE, our_fire.fire_alert_type)
```
Comment (LemonInTheDark) [inline code/datums/status_effects/debuffs/fire_stacks.dm:108]: You should use early returns here.

Comment (LemonInTheDark) [inline code/datums/status_effects/debuffs/fire_stacks.dm:108]: Rather then clearing the alert and on fire, then resetting it, it'd be better to have a var that's called "should_fire" or something. Defaults to FALSE, if it finds a source of fire, set it to true
Then use the owner's old on_fire var to decide it we should clear the alert, send a new one, or do nothing.

File: code/datums/status_effects/debuffs/fire_stacks.dm [hunk 19]
```diff
@@ -0,0 +1,373 @@
...
+	/// Current amount of stacks we have
+	var/stacks
+	/// Maximum of stacks that we could possibly get
+	var/stack_limit = 20
+	/// What status effect types do we remove uppon being applied. These are just deleted without any deduction from our or their stacks when forced.
+	var/list/enemy_types = list()
+	/// What status effect types do we merge into if they exist. Ignored when forced.
+	var/list/merge_types = list()
+	/// What status effect types do we override if they exist. These are simply deleted when forced.
+	var/list/override_types = list()
```
Comment (LemonInTheDark) [inline code/datums/status_effects/debuffs/fire_stacks.dm:16]: Is there any reason these need to be lists on init?
It adds a bit of extra cost to New, in general if it's not inconvienient it's best to avoid creating lists until you actually need them
for the record `for(var/thing in null)` will not runtime.

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/datums/status_effects/debuffs/fire_stacks.dm [hunk 20]
```diff
@@ -0,0 +1,376 @@
...
+		var/list/merge_effects = list()
+		for(var/merge_type in merge_types)
+			var/datum/status_effect/fire_handler/merge_effect = owner.has_status_effect(merge_type)
+			if(merge_effect)
+				merge_effects += merge_effects
+
+		if(LAZYLEN(merge_effects))
+			for(var/datum/status_effect/fire_handler/merge_effect in merge_effects)
+				merge_effect.adjust_stacks(stacks * stack_modifier / merge_effect.stack_modifier / LAZYLEN(merge_effects))
+			qdel(src)
```
Comment (LemonInTheDark) [inline code/datums/status_effects/debuffs/fire_stacks.dm:61]: return here, nothing should execute after a qdel(src)

File: code/datums/status_effects/debuffs/fire_stacks.dm [hunk 21]
```diff
@@ -0,0 +1,376 @@
...
+		var/no_protection = FALSE
+		if(victim.dna && victim.dna.species)
+			no_protection = victim.dna.species.handle_fire(victim, delta_time, times_fired, no_protection)
+		harm_human(delta_time, times_fired, no_protection)
+
+	if(isalien(owner))
+		owner.adjust_bodytemperature(BODYTEMP_HEATING_MAX * 0.5 * delta_time)
+
+	var/turf/location = get_turf(owner)
+	location.hotspot_expose(700, 25 * delta_time, TRUE)
```
Comment (LemonInTheDark) [inline code/datums/status_effects/debuffs/fire_stacks.dm:184]: I'm not a massive fan of this pattern, If you're going to handle subtypes like this I think inheritance is a better way of doing it

File: code/datums/status_effects/debuffs/fire_stacks.dm [hunk 22]
```diff
@@ -0,0 +1,376 @@
...
+		qdel(src)
+		return TRUE
+
+	if(!on_fire || isanimal(owner))
+		return TRUE
+
+	adjust_stacks(-0.05 * delta_time)
+
+	if(iscyborg(owner))
+		adjust_stacks(-0.5 * delta_time)
```
Comment (LemonInTheDark) [inline code/datums/status_effects/debuffs/fire_stacks.dm:147]: Kinda ditto here

File: code/datums/status_effects/debuffs/fire_stacks.dm [hunk 23]
```diff
@@ -0,0 +1,376 @@
...
+
+	if(thermal_protection >= FIRE_IMMUNITY_MAX_TEMP_PROTECT && !no_protection)
+		return
+
+	if(thermal_protection >= FIRE_SUIT_MAX_TEMP_PROTECT && !no_protection)
+		victim.adjust_bodytemperature(5.5 * delta_time)
+	else
+		victim.adjust_bodytemperature((BODYTEMP_HEATING_MAX + (stacks * 12)) * 0.5 * delta_time)
+		SEND_SIGNAL(victim, COMSIG_ADD_MOOD_EVENT, "on_fire", /datum/mood_event/on_fire)
+		victim.mind?.add_memory(MEMORY_FIRE, list(DETAIL_PROTAGONIST = victim), story_value = STORY_VALUE_OKAY)
```
Comment (LemonInTheDark) [inline code/datums/status_effects/debuffs/fire_stacks.dm:208]:
```suggestion
	if(no_protection)
		victim.adjust_bodytemperature((BODYTEMP_HEATING_MAX + (stacks * 12)) * 0.5 * delta_time)
		SEND_SIGNAL(victim, COMSIG_ADD_MOOD_EVENT, "on_fire", /datum/mood_event/on_fire)
		victim.mind?.add_memory(MEMORY_FIRE, list(DETAIL_PROTAGONIST = victim), story_value = STORY_VALUE_OKAY)
		return

	if(thermal_protection >= FIRE_IMMUNITY_MAX_TEMP_PROTECT)
		return
	if(thermal_protection >= FIRE_SUIT_MAX_TEMP_PROTECT)
		victim.adjust_bodytemperature(5.5 * delta_time)
```
Also for the record, I'd prefer no_protection be called protection. Negatives and double negatives can be hard to parse, best to just be as direct as possible to start, and then let the ! operator do the talking

File: code/datums/status_effects/debuffs/fire_stacks.dm [hunk 24]
```diff
@@ -0,0 +1,376 @@
...
+/**
+ * Handles clothing burning
+ *
+ * Arguments:
+ * - delta_time
+ * - times_fired
+ *
+ */
+
+/datum/status_effect/fire_handler/proc/burn_stuff(delta_time, times_fired)
```
Comment (LemonInTheDark) [inline code/datums/status_effects/debuffs/fire_stacks.dm:219]: This also feels like it should be a proc on carbon or something

File: code/datums/status_effects/debuffs/fire_stacks.dm [hunk 25]
```diff
@@ -0,0 +1,376 @@
...
+			victim.apply_overlay(FIRE_LAYER)
+			last_icon_state = fire_icon
+
+		else if(last_icon_state)
+			victim.remove_overlay(FIRE_LAYER)
+			victim.apply_overlay(FIRE_LAYER)
+			last_icon_state = null
+
+		return TRUE
+	return FALSE
```
Comment (LemonInTheDark) [inline code/datums/status_effects/debuffs/fire_stacks.dm:354]: Yeeesh

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/datums/status_effects/debuffs/fire_stacks.dm [hunk 26]
```diff
@@ -0,0 +1,379 @@
...
+		victim.overlays_standing[FIRE_LAYER] = new_fire_overlay
+		victim.apply_overlay(FIRE_LAYER)
+		last_icon_state = fire_icon
+
+	else if(last_icon_state)
+		victim.remove_overlay(FIRE_LAYER)
+		victim.apply_overlay(FIRE_LAYER)
+		last_icon_state = null
+
+	return TRUE
```
Comment (LemonInTheDark) [inline code/datums/status_effects/debuffs/fire_stacks.dm:357]: I still have a problem with this
If you're going to be sticking istypes in this proc, it should be handled instead by proc overrides
the same is true of a lot of this code

that doesn't mean you need to drop the modularity, I just want the override logic handled by overrides, and not interwoven istypes

File: code/datums/status_effects/debuffs/fire_stacks.dm (hunk 21 above)
Comment (LemonInTheDark) [reply code/datums/status_effects/debuffs/fire_stacks.dm:185]: I still feel this way

File: code/modules/mob/living/living.dm [hunk 27]
```diff
@@ -1445,38 +1445,75 @@
...
+
+/**
+ * Sets fire overlay of the mob.
+ *
+ * Vars:
+ * * stacks: Current amount of fire_stacks
+ * * on_fire: If we're lit on fire
+ * * last_icon_state: Holds last fire overlay icon state, used for optimization
+ * * suffix: Suffix for the fire icon state for special fire types
+ */
```
Comment (LemonInTheDark) [inline code/modules/mob/living/living.dm:1514]: document the return

File: code/modules/mob/living/living.dm [hunk 28]
```diff
@@ -1445,38 +1445,75 @@
...
+
+/mob/living/proc/update_fire_overlay(stacks, on_fire, last_icon_state, suffix = "")
```
Comment (LemonInTheDark) [inline code/modules/mob/living/living.dm:1517]:
```suggestion
/mob/living/proc/update_fire_overlay(stacks, on_fire, last_icon_state, suffix = "")
	return
```

File: code/datums/status_effects/debuffs/fire_stacks.dm [hunk 29]
```diff
@@ -0,0 +1,273 @@
...
+		var/mob/living/carbon/human/victim = owner
+		SEND_SIGNAL(victim, COMSIG_HUMAN_BURNING)
+		victim.burn_clothing(delta_time, times_fired, stacks)
+		var/no_protection = FALSE
+		if(victim.dna && victim.dna.species)
+			no_protection = victim.dna.species.handle_fire(victim, delta_time, times_fired, no_protection)
+		harm_human(delta_time, times_fired, no_protection)
+
+	if(isalien(owner))
+		owner.adjust_bodytemperature(BODYTEMP_HEATING_MAX * 0.5 * delta_time)
```
Comment (LemonInTheDark) [inline code/datums/status_effects/debuffs/fire_stacks.dm:182]: I feel semi similar about these, but less strongly

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/datums/status_effects/debuffs/fire_stacks.dm [hunk 30]
```diff
@@ -326,34 +326,35 @@
...
+			fire_icon = "[human_victim.dna.species.fire_overlay][get_special_icon()]"
+		else
+			fire_icon = "human_burning[get_special_icon()]"
+
+	if((stacks > 0 && on_fire) || HAS_TRAIT(victim, TRAIT_PERMANENTLY_ONFIRE))
+		if(fire_icon == last_icon_state)
+			return TRUE
+
+		victim.remove_overlay(FIRE_LAYER)
+		var/mutable_appearance/new_fire_overlay = mutable_appearance('icons/mob/onfire.dmi', fire_icon, -FIRE_LAYER)
```
Comment (Kapu1178) [inline code/datums/status_effects/debuffs/fire_stacks.dm]: this should be cached i m o

Comment (LemonInTheDark) [reply code/datums/status_effects/debuffs/fire_stacks.dm]: This is not a bad point

File: code/datums/status_effects/debuffs/fire_stacks.dm [hunk 31]
```diff
@@ -0,0 +1,263 @@
...
+		if(override_effect)
+			if(forced)
+				qdel(override_effect)
+				continue
+
+			adjust_stacks(override_effect.stacks)
+			qdel(override_effect)
+
+	. = ..()
+	cache_stacks()
```
Comment (LemonInTheDark) [inline code/datums/status_effects/debuffs/fire_stacks.dm:75]: why?

File: code/datums/status_effects/debuffs/fire_stacks.dm [hunk 32]
```diff
@@ -0,0 +1,263 @@
...
+
+/**
+ * Handles mob extinguishing, should be the only way to set on_fire to FALSE
+ */
+
+/datum/status_effect/fire_handler/fire_stacks/proc/extinguish()
+	if(firelight_ref)
+		var/atom/firelight = firelight_ref.resolve()
+		if(firelight && !QDELETED(firelight))
+			qdel(firelight)
```
Comment (LemonInTheDark) [inline code/datums/status_effects/debuffs/fire_stacks.dm:233]: You can just qdel the reference and it'll do this for ya 
image

File: code/datums/status_effects/debuffs/fire_stacks.dm [hunk 33]
```diff
@@ -0,0 +1,263 @@
...
+
+	on_fire = FALSE
+	SEND_SIGNAL(owner, COMSIG_CLEAR_MOOD_EVENT, "on_fire")
+	SEND_SIGNAL(owner, COMSIG_LIVING_EXTINGUISHED, owner)
```
Comment (LemonInTheDark) [inline code/datums/status_effects/debuffs/fire_stacks.dm:237]: God I fucking FLOYYYYYYYYYD

File: code/modules/mob/living/carbon/carbon.dm [hunk 34]
```diff
@@ -1424,3 +1414,24 @@
...
+		var/mutable_appearance/new_fire_overlay = mutable_appearance('icons/mob/onfire.dmi', fire_icon, -FIRE_LAYER)
+		new_fire_overlay.appearance_flags = RESET_COLOR
+		overlays_standing[FIRE_LAYER] = new_fire_overlay
+		apply_overlay(FIRE_LAYER)
+		last_icon_state = fire_icon
+
+	else if(last_icon_state)
+		remove_overlay(FIRE_LAYER)
+		apply_overlay(FIRE_LAYER)
+		last_icon_state = null
```
Comment (LemonInTheDark) [inline code/modules/mob/living/carbon/carbon.dm:1435]: Use early returns

File: code/modules/mob/living/carbon/carbon.dm [hunk 35]
```diff
@@ -1424,3 +1414,24 @@
...
+
+/mob/living/carbon/update_fire_overlay(stacks, on_fire, last_icon_state, suffix = "")
+	var/fire_icon = "generic_burning[suffix]"
+
+	if((stacks > 0 && on_fire) || HAS_TRAIT(src, TRAIT_PERMANENTLY_ONFIRE))
+		if(fire_icon == last_icon_state)
+			return last_icon_state
+
+		remove_overlay(FIRE_LAYER)
+		var/mutable_appearance/new_fire_overlay = mutable_appearance('icons/mob/onfire.dmi', fire_icon, -FIRE_LAYER)
```
Comment (LemonInTheDark) [inline code/modules/mob/living/carbon/carbon.dm:1426]: any reason this can't be cached?

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 36]
```diff
@@ -966,3 +966,94 @@
...
+		overlays_standing[FIRE_LAYER] = new_fire_overlay
+		apply_overlay(FIRE_LAYER)
+		last_icon_state = fire_icon
+
+	else if(last_icon_state)
+		remove_overlay(FIRE_LAYER)
+		apply_overlay(FIRE_LAYER)
+		last_icon_state = null
+
+	return last_icon_state
```
Comment (LemonInTheDark) [inline code/modules/mob/living/carbon/human/human_defense.dm:1051]: ditto here for both

File: code/modules/mob/living/living.dm [hunk 37]
```diff
@@ -1445,38 +1445,90 @@
...
+			spread_to.set_fire_stacks(firesplit, fire_type)
+		else
+			adjust_fire_stacks(-fire_stacks / 2, fire_status.type)
+			spread_to.adjust_fire_stacks(fire_stacks, fire_status.type)
+			if(spread_to.ignite_mob())
+				log_game("[key_name(src)] bumped into [key_name(spread_to)] and set them on fire")
+	else if(their_fire_status && their_fire_status.on_fire)
+		spread_to.adjust_fire_stacks(-spread_to.fire_stacks / 2, their_fire_status.type)
+		adjust_fire_stacks(spread_to.fire_stacks, their_fire_status.type)
+		ignite_mob()
```
Comment (LemonInTheDark) [inline code/modules/mob/living/living.dm:1504]: early returns please

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/datums/status_effects/debuffs/fire_stacks.dm [hunk 38]
```diff
@@ -0,0 +1,261 @@
...
+		if(override_effect)
+			if(forced)
+				qdel(override_effect)
+				continue
+
+			adjust_stacks(override_effect.stacks)
+			qdel(override_effect)
+
+	. = ..()
+	cache_stacks()
```
Comment (LemonInTheDark) [inline code/datums/status_effects/debuffs/fire_stacks.dm:75]: I am still kinda miffed that this is required. I don't think the cost of calling parent here earlier is that high, and it would remove a class of error from people working with your system. Is there a strong reason to not do it?

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
